server/blob/blob-api/src/main/java/org/apache/james/blob/api/Store.java

```diff
@@ -0,0 +1,114 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.api;
+
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.util.FluentFutureStream;
+
+import com.google.common.collect.ImmutableMap;
+
+public interface Store<T> {
+    class BlobType {
+        private final String name;
+
+        public BlobType(String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof BlobType) {
+                BlobType blobType = (BlobType) o;
+
+                return Objects.equals(this.name, blobType.name);
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(name);
+        }
+    }
+
+    interface Encoder<T> {
+        Stream<Pair<BlobType, InputStream>> encode(T t);
+    }
+
+    interface Decoder<T> {
+        void validateInput(Collection<BlobType> input);
+
+        T decode(Stream<Pair<BlobType, byte[]>> streams);
+    }
+
+    CompletableFuture<Map<BlobType, BlobId>> save(T t);
+
+    CompletableFuture<T> read(Map<BlobType, BlobId> blobIds);
```

Comment:
As a caller, if I want to retrieve my Object from blob store. I'm expecting I just need only the blobId instead of both blobType and blobId. E.g. if I want to retrieve an user from the database, I just only need the userId.

Reply:
I did some works with generics to allow the basic interface to return any kind of Id.

The I,pl enforce a structure of the Id.